Markdown parser for translation notes and questions

KJoslyn opened the pull request:
Markdown parser to parse translation notes and questions from .md files contained in the following two RC locations:
https://git.door43.org/WycliffeAssociates/en_tn
https://git.door43.org/WycliffeAssociates/en_tq

For example, view the raw data for the file:
https://git.door43.org/WycliffeAssociates/en_tn/src/branch/master/1ti/02/11.md

Notes and questions are parsed the same way. The title text (either the note "snippet" or the question) are parsed by locating one or more '#' characters at the start of the line of text (it appears that all .md files for verses only have one '#' character to denote a new resource.) The body text (either the note about the snippet or the answer to the question) are parsed by reading text from subsequent lines. The title and body are placed into a HelpResource data object. The output of the parser's parse() function is a list of these HelpResource objects.

Currently, the parser does not remove any references or links at the end of the body text. This is due to the assumption that the narrator will not read this part, and something may be done with these references later.

Additionally, the parser only truly supports .md files that are for verses, and not for intro.md files for chapters or books (for example, https://git.door43.org/WycliffeAssociates/en_tn/src/branch/master/1ti/02/intro.md)

These intro files may have a hierarchy of headings. Additionally, they may have several lines of body text, which may be separated by blank lines. (It is unknown whether any .md files for verses have multiple lines of body text.) Hierarchies of headings are currently handled by creating HelpResource data objects for each heading which may have empty body text. Multiple lines of body text are currently supported by concatenating the lines with spaces between the lines.

In the future, we may need to add functionality to the parser to support intro.md files. Other known issues for intro.md files are '#' symbols at the end of the title text, and a dedicated section for links.

Finally, this PR includes the addition of method stubs for missing MockResourceMetadataRepository overrides. While unrelated to the parser, it appears these method stubs were missing and this was preventing the build of the project.

KJoslyn committed:
Moved usfm package to new 'parse' package and created empty IParser interface

KJoslyn committed:
Revert "Moved usfm package to new 'parse' package and created empty IParser interface"

This reverts commit ddd6f46101c18f893429af7592b9853056747676.

KJoslyn committed:
Created ParseMd.kt for tn, tq

KJoslyn committed:
Created ParseMdTest.kt with one unit test

KJoslyn committed:
Added robustness to parser and created more test functions and cases

KJoslyn committed:
Added method stubs for missing MockResourceMetadataRepository overrides

aunger commented on src/test/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMdTest.kt:
```diff
@@ -0,0 +1,175 @@
+    // (including the branching logic of the parse() function)
+    private val testParseCases = listOf(
+            mapOf(
+                    "lines" to listOf(
```
Not worth changing, but multi-line string literals might be tidier here: https://kotlinlang.org/docs/reference/basic-types.html#string-literals

aunger commented on src/test/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMdTest.kt:
```diff
@@ -0,0 +1,175 @@
+                            "",
+                            "Body 2"
+                    ),
+                    "expected" to arrayListOf(
```
Could be listOf.  Below also.

KJoslyn replied:
Not sure why I did that when I used listOf above, thanks!

aunger commented on src/test/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMdTest.kt:
```diff
@@ -0,0 +1,175 @@
+                    "line" to "# ",
+                    "expected" to false
+            ),
+            mapOf(
```
You're using `Map<String, String>` as an ad-hoc way to make dynamically typed objects. It works here since this is a small source file, but you lose type safety and null safety. All the `as String`s in the code below are a sign that you're working against Kotlin's null safety.

As an example, if you mistype "expetced" as your key, you'll get back a null String, which you then cast to String.  Like I said, it really isn't causing a problem because this is such a small file and any bug here is just going to make a test case fail.  NBD, and I know I'm being picky, but the first PR is a good time to talk about this stuff. 

Please experiment with converting all your `Map`s to data classes with `input` and `expected` fields.  You can make a single generic data class that handles either String or Bool types for `expected`, or you can just make two classes.

Or, perhaps better and simpler, since all of your classes really only have those two fields, you could just use [`to`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/to.html) to create input-to-expected pairs.

KJoslyn replied:
Simply using to is a much better approach, thanks for the suggestion!

aunger commented on src/test/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMdTest.kt:
```diff
@@ -0,0 +1,175 @@
+
+        val stream: ByteArrayInputStream = lines.joinToString(lineSeparator).byteInputStream()
+
+        return BufferedReader(stream.bufferedReader())
```
I think using the BR constructor is redundant here.

KJoslyn replied:
Good catch, maybe this was an artifact of an earlier failed attempt. This is one of those duh moments!

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMd.kt:
```diff
@@ -0,0 +1,56 @@
+// TODO: Add Help type enum to HelpResource? (tn, tq)
+data class HelpResource(var title: String, var body: String)
+
+class HelpResourceList: ArrayList<HelpResource>()
```
This should be of type `List` instead of `ArrayList`, so you're not locked into this implementation. Even if you want to build it using a mutable list such as `ArrayList`, it should be presented to consumers as `List`.

jsarabia replied:
Note that if you wish to return a list that has the ability to add elements, the List interface is immutable in kotlin, unlike Java, and you'd need to do MutableList instead.

aunger commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMd.kt:
```diff
@@ -0,0 +1,56 @@
+// TODO: Add Help type enum to HelpResource? (tn, tq)
+data class HelpResource(var title: String, var body: String)
+
+class HelpResourceList: ArrayList<HelpResource>()
```
[Type aliases](https://kotlinlang.org/docs/reference/type-aliases.html) might be better than subclassing here.  I'm new to Kotlin, of course, so I don't know.  It may be best just not to bother renaming to `HelpResourceList`, and just use `List<HelpResource>`.

jsarabia replied:
Please use List<HelpResource> (or MutableList<HelpResource> if that's needed), it's consistent with how we use lists everywhere else in the codebase.

I hadn't seen type aliasing before, so thanks for that!

KJoslyn replied:
Type aliases looks like the way to go here compared to subclassing. But now I think it might be better not to even bother creating HelpResourceList, since List<HelpResource> is actually more straightforward and not very complicated.

aunger reviewed (changes requested):
Looks good! My comments are all minor.

KJoslyn committed:
Addressing review comments

natibekele commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMd.kt:
```diff
@@ -0,0 +1,54 @@
+        private val isTitleRegex = Regex("^#+\\s*[^#\\s]+")
+        private val titleTextRegex = Regex("^#+\\s*")
+
+        fun parse(reader: BufferedReader): List<HelpResource> {
```
I'm just being picky too, and curious, but wouldn't it be better if the parser takes in the `File` that it needs to parse as opposed to a `BufferedReader`? In practice I suppose I wouldn't want to have to set up the reader every time i wanted to parse some MD. I'd think the reader would be contained in the parser and i wouldn't have to worry about that

jsarabia replied:
A BufferedReader can be created from an InputStream which is the preferred way of accessing resources files, streaming a file from a url without saving it to the filesystem first, etc. File would add restrictions to how the parser could be used, so I'm good with keeping it a BufferedReader.

Good question!

KJoslyn replied:
@natibekele @sarabiaj This was actually a result of a conversation with Craig about unit testing. If the parser takes in a file, the unit test also has to supply a file, which is problematic since 1. it takes longer during unit test to read from a file, and 2. we don't want our unit tests to be dependent on files in specific directories. Thought I'd share this, since I thought Craig made some really good points!

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMd.kt:
```diff
@@ -0,0 +1,54 @@
+            return helpResourceList
+        }
+
+        internal fun getTitleText(line: String): String {
```
These two functions could be inlined. Not a big deal, but a free (admittedly trivial) performance increase lol.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMd.kt:
```diff
@@ -0,0 +1,54 @@
+
+        internal fun getTitleText(line: String): String {
+
+            return line.removePrefix(titleTextRegex.find(line)!!.value)
```
I know you check first to see if the title matches before you call this function, so you WILL match here; but it would still be better to refactor this to avoid the null, in the event that in the future this function is used and the if check is not done before calling it.

jsarabia commented on src/main/kotlin/org/wycliffeassociates/otter/common/domain/md/ParseMd.kt:
```diff
@@ -0,0 +1,54 @@
+
+            reader.forEachLine {
+
+                if (it.isEmpty())
```
I'd rather always have { } around an if block even if it is one line. 

As an alternative, you could make it one line:
```
if (it.isEmpty()) return@forEachLine
```

jsarabia reviewed (changes requested):
I'd prefer a more conservative use of whitespace. I notice you put a new line after every if, method signature, for, etc, but I would prefer not doing that. 

New lines between a block of variables and the beginning of method code are good, lines to separate groups of related variables, lines to separate complete logical thoughts are good.

KJoslyn committed:
Addressing front matter meeting discussion and review comments

From front matter meeting:
1. Hashtags are now kept at the start of the titles so that they can be rendered or stripped as desired for display
2. Multiple lines of body text are now separated by newlines

Also addressed recent review comments such as whitespace and !! assertion.

KJoslyn committed:
Made ParseMd an object
